File: Graph/Topological sort.py
    '''
    https://practice.geeksforgeeks.org/problems/topological-sort/1#
    '''
  
  #TC:O(V+E)
    #SC:O(V)
    #Function to return list containing vertices in Topological order.
    
    #Refer William Fiset Video
    
    #Algorithm
    #1.Pick an unvisited node
    #2.After selecting a node,mark it as visited and perform DFS on its adjacent unvisited nodes only
    #3.Continue this till we get to a node having no outward edges.
    #4.On the recursive callback add the current node to the ordering list in reverse order.
    
    def topoSort(self, V, adj):
        visited = [False]*V
        #This list keeps all my nodes in its topological sorted order
        ordering = [-1]*V
        i = V-1
        #As we will be using DFS,therefore we will store the nodes in a reverse fashion
        #in my ordering list
        for node in range(V):
            if not visited[node]:
              i = self.DFS(node,i,ordering,adj,visited)
        #If one subset of graph is visited than it returns the index of the ordering list
        #from which we should start inserting the node of other subsets of the graph
        return ordering
        
        #i denotes the index at which we should insert the node value.
    def DFS(self,node,i,ordering,adj,visited):
        visited[node]=True
        for neighbor in adj[node]:
            if not visited[neighbor]:
               i =  self.DFS(neighbor,i,ordering,adj,visited)
        ordering[i] = node
        return i-1

File: Graph/Topological sort.java
    '''
    https://practice.geeksforgeeks.org/problems/topological-sort/1#
    '''
//Algo:
/*
Algorithm: Steps involved in finding the topological ordering of a DAG:
Step-1: Compute in-degree (number of incoming edges) for each of the vertex present in the DAG and initialize the count of visited nodes as 0.

Step-2: Pick all the vertices with in-degree as 0 and add them into a queue (Enqueue operation)

Step-3: Remove a vertex from the queue (Dequeue operation) and then.
Increment count of visited nodes by 1.
Decrease in-degree by 1 for all its neighboring nodes.
If in-degree of a neighboring nodes is reduced to zero, then add it to the queue.
Step 5: Repeat Step 3 until the queue is empty.
Step 6: If count of visited nodes is not equal to the number of nodes in the graph then the topological sort is not possible for the given graph.
*/
  #TC:O(V+E)
    #SC:O(V)
    #Function to return list containing vertices in Topological order.
    
class Solution
{
    //Function to return list containing vertices in Topological order. 
    static int[] topoSort(int V, ArrayList<ArrayList<Integer>> adj) 
    {
        Queue<Integer> q = new LinkedList<Integer>();
        int[] indegrees = new int[V];
        ArrayList<Integer> topOrder = new ArrayList<>();
        
        int noOfNodes = 0;
        
        for(int v=0;v<adj.size();v++){
            for(int adjacent:adj.get(v)){
                indegrees[adjacent]+=1;
            }
        }
        
        for(int i=0;i<V;i++){
                if(indegrees[i]==0){
                    q.add(i);
                }
            }
        while(!q.isEmpty()){
            
            int vertex = q.poll();
            topOrder.add(vertex);
            noOfNodes++;
            for(int adjacent:adj.get(vertex)){
                if(--indegrees[adjacent]==0){
                    q.add(adjacent);
                    }
                }
            }
        
    int[] arr = new int[topOrder.size()];
    for(int i=0;i<arr.length;i++){
        arr[i] = topOrder.get(i);
    }
    return arr;
    }
}
